encrypt-config/ArgParser.cpp

```diff
@@ -0,0 +1,178 @@
+    ss << " ";
+    if (!simple_arg.required) {
+      ss << "[";
+    }
+    ss << utils::StringUtils::join("|", simple_arg.names)
+        << " <" << simple_arg.value_name << ">";
+    if (!simple_arg.required) {
+      ss << "]";
+    }
+  }
+  for (const auto& flag : flag_arguments_) {
+    ss << " [" << utils::StringUtils::join("|", flag.names) << "]";
+  }
+  ss << "\n";
+  for (const auto& simple_arg : simple_arguments_) {
+    ss << "\t";
+    ss << utils::StringUtils::join("|", simple_arg.names) << " : ";
+    if (simple_arg.required) {
+      ss << "(required)";
+    } else {
+      ss << "(optional)";
+    }
+    ss << " " << simple_arg.description;
+    ss << "\n";
+  }
+  for (const auto& flag : flag_arguments_) {
+    ss << "\t" << utils::StringUtils::join("|", flag.names) << " : "
+        << flag.description << "\n";
+  }
+  return ss.str();
+}
+
+void Arguments::set(const std::string& key, const std::string& value) {
+  if (get(key)) {
+    throw CommandException("Key is specified more than once \"" + key + "\"");
+  }
+  simple_args_[key] = value;
+}
+
+void Arguments::set(const std::string& flag) {
+  if (isSet(flag)) {
+    throw CommandException("Flag is specified more than once \"" + flag + "\"");
+  }
+  flag_args_.insert(flag);
+}
+
+utils::optional<std::string> Arguments::get(const std::string &key) const {
+  return getSimpleArg(key) | utils::flatMap([&] (const Argument& arg) {return get(arg);});
+}
+
+utils::optional<std::string> Arguments::get(const Argument& arg) const {
+  for (const auto& name : arg.names) {
+    auto it = simple_args_.find(name);
+    if (it != simple_args_.end()) {
+      return it->second;
+    }
+  }
+  return {};
+}
+
+bool Arguments::isSet(const std::string &flag) const {
+  utils::optional<FlagArgument> opt_flag = getFlag(flag);
+  if (!opt_flag) {
+    return false;
+  }
+  return haveCommonItem(opt_flag->names, flag_args_);
+}
+
+Arguments Arguments::parse(int argc, char* argv[]) {
+  Arguments args;
+  for (int argIdx = 1; argIdx < argc; ++argIdx) {
+    std::string key{argv[argIdx]};
+    if (getFlag(key)) {
+      args.set(key);
+      continue;
+    }
+    if (!getSimpleArg(key)) {
+      throw CommandException("Unrecognized option: \"" + key + "\"");
+    }
+    if (argIdx == argc - 1) {
+      throw CommandException("No value specified for key \"" + key + "\"");
+    }
+    ++argIdx;
+    std::string value{argv[argIdx]};
+    args.set(key, value);
+  }
+  if (args.isSet("-h")) {
+    std::cout << getHelp();
+    std::exit(0);
+  }
+  for (const auto& simple_arg : simple_arguments_) {
+    if (simple_arg.required && !args.get(simple_arg)) {
+      throw CommandException("Missing required option " + utils::StringUtils::join("|", simple_arg.names));
+    }
+  }
+  return args;
+}
+
+utils::optional<FlagArgument> Arguments::getFlag(const std::string &name) {
+  for (const auto& flag : flag_arguments_) {
+    if (flag.names.count(name) > 0) {
```

Comment:
I favor comparison over implicit integer -> boolean conversion. This one is also more expressive IMO, because the question is whether the key exists and the closes phrasing with count is "if occurrence count is greater than zero" (rather than "if occurrence count")

Reply:
I mean on a container with unique elements `count()` gives you either `0` or `1` and is used for exactly this purpose. We can explicitly say that we want to convert the type returned from a yes or no interface to a boolean yes or no interface, but IMO it is redundant.